### ISSUE TITLE
Unable to connect using SSL/TLS

### ISSUE THREAD

**Issue by drschwabe** [sstur/nodeftpd#54]
Is anyone else hosting their app over SSL/TLS ? 

I have an SSL cert specifically for "ftp.mydomain.com" which is pointed at my application running nodeftp but have had no luck getting it to work. 

The output of my error is as follows: 

```
FTPD listening on port 21
Client connected from xx.xx.xxx.xxx
Establishing secure connection...
Secure socket not authorized: [Error: UNABLE_TO_GET_ISSUER_CERT]
```

I've tried loading the intermediaries into one .pem file as suggested [here](https://github.com/coolaj86/node-ssl-root-cas) as well as just passing them individually in the ca array - but to no avail.

Here's the code. In this example my app uses a Comodo PositiveSSL cert.  

```
//For SSL/TLS:
var tlsOptions = {
  key: fs.readFileSync('ftp_mydomain_com.key'),
  cert: fs.readFileSync('ftp_mydomain_com.crt'),
  ca : [
    fs.readFileSync('ftp_mydomain_com.crt'),  
    fs.readFileSync('COMODORSADomainValidationSecureServerCA.crt'),
    fs.readFileSync('COMODORSAAddTrustCA.crt'), 
    fs.readFileSync('AddTrustExternalCARoot.crt')
  ],
  rejectUnauthorized: true
}

var options = {
  tlsOptions: tlsOptions,
  allowUnauthorizedTls: false,
  tlsOnly: true,
  getInitialCwd: function(connection, initialCwdCallback) {
  //and so forth
```
   
The user can connect if if you set "allowUnauthorizedTls".
Also, if you disable "tlsOnly" plain FTP works fine too.

**Comment by sstur** [comment]
Hi @drschwabe I don't use the SSL/TLS features of nodeftpd so sorry I can't be much help here. In fact I've been thinking about pulling the crypto out of this project since FTP is poorly suited for encryption and I feel like there are better options for securely transferring files (SSH or maybe HTTPS).

In your case, what if you try another FTP server with your same certs, maybe vsftpd does that work? If so then it is probably a bug in our implementation.

**Comment by addrummond** [comment]
Hi @drschwabe, I put the SSL code into nodeftpd originally but I have to admit that it's a while since I tried using it. When I wrote it, there wasn't any "official" way of upgrading an insecure connection. The code in nodeftpd is based on this:

https://github.com/andris9/rai/commits/master/lib/starttls.js

I think it's been modified a bit since I incorporated it into nodeftpd. It might be worth seeing if you can get that code to work outside of nodeftpd. I'll also try to find out if there's an official API for doing this now.

**Comment by drschwabe** [comment]
Okay, good ideas, thanks for the tips!  Will give vsftpd a try to help rule out any potential issue with my certs.   And maybe tinker with rai.

**Comment by boydlee** [comment]
Actually having the same problems, did you have any luck with this @drschwabe?

**Comment by addrummond** [comment]
Since tls support was added to nodeftpd, the code in starttls.js has been packaged up into an npm module with a somewhat different API.

https://www.npmjs.com/package/starttls

I'll see if using this module instead of the older code helps with the cert issue.

**Comment by johnculviner** [comment]
@drschwabe @boydlee Anyone figured out a way to get this to work? Need TLS unfortunately here too...

**Comment by drschwabe** [comment]
@boydlee @johnculviner Haven't gotten it to work, so intend on going with vsftpd and will build my node application around that.

**Comment by boydlee** [comment]
I ended up going the same route except using pureftp and a mysql db to
manage the accounts dynamically.

**Comment by johnculviner** [comment]
Thanks for the update folks! I was thinking a way to get around it might be to terminate SSL on nginx or something too. I'll update if I tink of anything

**Comment by matiaslarsson** [comment]
+1

**Comment by johnculviner** [comment]
hey @matiaslarsson  this does work!

My problem was that my private key and public cert weren't correct for node. This works for me
```
var serverOptions = {
	allowUnauthorizedTls: true,
	tlsOnly: true,
	tlsOptions: {
		key: fs.readFileSync('private-key.pem'),
		cert: fs.readFileSync('public-cert.pem')
	}
};

var server = new ftpd.FtpServer('localhost', serverOptions);
//....
```
if the below doesn't work for you make sure the certs you are using work with a simple Node HTTPS server. If they don't then you know thats the culprit. You get it working with a simple Node HTTPS server the above should work. (This was like a half a day of my life so hopefully this helps!)

**Comment by matiaslarsson** [comment]
Thanks for the info @johnculviner! I have implemented the beginning of an Azure Blob Storage "file system" for nodeftpd and FTPS would be awesome.

**Comment by sstur** [comment]
@matiaslarsson: That seems like a cool project. I hope nodeftpd is helpful to you. I do believe, however, that FTP is a poorly suited protocol for secure transfers, even when adding TLS. In fact, the best way to do FTP securely, I think is to run an FTP server on localhost that really communicates with a remote file system via a much more modern, secure protocol. This is how [shopify-ftp](https://github.com/sstur/shopify-ftp) works and I believe one could do the same thing for other protocols like AWS S3, Amazon Cloud Drive, Azure Blob Storage, etc. For this reason, I'd like to take TLS out of this library completely and deprecate it's use. I would love to see the industry come together around a secure, open protocol for remote file transfer and see FTP become a thin compatibility layer for use with legacy clients. Maybe that's too much to ask. But the fact remains that Secure FTP is a bit of an anti-pattern.

**Comment by johnculviner** [comment]
Totally agree that FTP is silly, TLS over FTP is silly etc. but sometimes for interop purposes FTP is the only option. What was awesome about this project is I could stand up a virtual FTP server that actually does interact with a different backend (its NOT the filesystem but actually a database) but to our clients that love FTP, well.. its 'FTP'. Please don't remove TLS support, working fine for me!

**Comment by sstur** [comment]
OK, if people are using this feature, I won't remove it. If I need to remove it to refactor the project, I'll fork it or work on another branch. Thanks!

**Comment by matiaslarsson** [comment]
I agree @sstur but as @johnculviner mentioned there are so many legacy systems/services out there still using the old trusty FTP-protocol.

Thanks for not removing it! :)

**Comment by adamdunkley** [comment]
I know this is way after the fact but we have a use case that requires this too (we connect with a very wide range of clients many of whom exclusively use FTP). We can't change that fact but we can do everything in our power to protect the communication, this means running TLS over FTP unfortunately.

I tried several approaches to fixing the problem (such as using a proxy, terminating the SSL and using implicit TLS) but, ultimately, the native option seems like the best (/only) way to proceed.

@sstur Just to let you know I have fixed a bug in the TLS server where it wasn't forwarding the 'finished' event. This is used to determine the end of the write stream so I began to see uploads hang when I switched on TLS. Let me know if there's anything I can do to help get this merged down https://github.com/sstur/nodeftpd/pull/118

Broadly, this project is really important to me and my team here at near.st so I'd be happy to discuss how we can contribute to the project to ensure it continues to work well.
